### PR TITLE
Remove Dbreeze async wrappers

### PR DESCRIPTION
**Pull request by noescape00**
Removed Dbreeze async wrappers to reduce overhead

ps
don't merge yet, posting it as a PR to make sure it passes on CI

### CODE REVIEW

**Comment by dangershony** [comment]
I will pull this PR and run mainnet

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs [hunk 1]
```diff
@@ -202,11 +202,11 @@ public async Task<GetTxOutModel> GetTxOutAsync(string txid, uint vout, bool incl
             UnspentOutputs unspentOutputs = null;
             if (includeMemPool)
             {
-                unspentOutputs = this.pooledGetUnspentTransaction != null ? await this.pooledGetUnspentTransaction.GetUnspentTransactionAsync(trxid) : null;
+                unspentOutputs = this.pooledGetUnspentTransaction != null ? await this.pooledGetUnspentTransaction.GetUnspentTransactionAsync(trxid).ConfigureAwait(false) : null;
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: Why remove the `ConfigureAwait` ?

**Comment by dangershony** [comment]
Looks good to me, once I sync networks this can go in

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs (hunk 1 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: Forgot to reintroduce it after making GetUnspentTransactionAsync not async and then changing my mind and putting it back.

Comment (noescape00) [reply src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: It was added in this PR actually

**Comment by MithrilMan** [comment]
What's the reason behind this changes?
What kind of overhead did you have? Not having await now all these are blocking calls, do you have evidence of performance improvements?

If it's about DBreeze not having Async, the problem is that these changes affect any kind of new db we may use, because they change interfaces and base classes definition and not just classes where db calls are actually done

**Comment by noescape00** [comment]
>What's the reason behind this changes?

- Reduce amount of code lines without any logic impact
- Do not wrap not async code into a task, it gives no benefit
- Reason why we kept async wrappers became obsolete 


>because they change interfaces and base classes definition 

Yes. 
Let's say method A calls B, B calls C, C calls D
If D contains any code that can be awaited all A,B,C,D should become async. 

If none of them contain any async code there is no reason for them to be async.

**Comment by MithrilMan** [comment]
Yes but the problem I see is that if we implement the storage layer with a storage implementation that supports async, we can't use it because the caller doesn't have async support

**Comment by MithrilMan** [comment]
> Yes but the problem I see is that if we implement the storage layer with a storage implementation that supports async, we can't use it because the caller doesn't have async support and you'd have to revert these changes again to support it.

And even more, we don't know if a 3rd party would come out with its own persistence layer that's async. Personally I'd leave async chain there

**Comment by noescape00** [comment]
>Yes but the problem I see is that if we implement the storage layer with a storage implementation that supports async, we can't use it because the caller doesn't have async support

That's the reason why we kept them before. But it seems to me there is no db suitable for us that supports async. There are leveldb and dbreeze, that's it. 
Looks like we will be sticking with one of those in any case.

**Comment by noescape00** [comment]
>Personally I'd leave async chain there

We can always reintroduce it. 
There is no reason to keep redundancies in the code. There is a small chance they will be needed in the future and if that happens it's trivial to reintroduce async support.

**Comment by MithrilMan** [comment]
anyway If you think this changes give us a big benefit then merge it

**Comment by noescape00** [comment]
>anyway If you think this changes give us a big benefit then merge it

No, not a big one. But I think it's worth it and benefits are greater than potential benefits of keeping the wrappers.

**Comment by dangershony** [comment]
@MithrilMan I don't see how an embedded db will be async, if db is not serverless for sure async makes sense, to have an async db inprocesses is less likely do you have examples of async embedded?

**Comment by MithrilMan** [comment]
Azure?

**Comment by MithrilMan** [comment]
btw if our calls start as async but db doesn't have awaits, it will run as synchronous and AFAIR it should  short circuit the awaiter, so shouldn't have noticeable performance hit (no context capture, etc...)

**Comment by dangershony** [comment]
>it could be an Azure storage.
No for consensus storage it would always be embeded.

**Comment by rowandh** [comment]
Looks like the main change here is not to remove Dbreeze async wrappers. It's to make `DBreezeCoinViewRule` operations synchronous by not running them on a new thread using `Task.Run`.

Because of this, returning `Task` becomes unnecessary and so removing async wrappers is the remainder of the changes.

So the main question here is - do we want to make `DBreezeCoinViewRule` operations synchronous?

**Comment by noescape00** [comment]
>So the main question here is - do we want to make DBreezeCoinViewRule operations synchronous?

They will be synchronous anyway. 
If we've used coinview like that: 

```
var task = coinview.DoSmthAsync()

do something heavy that can be done in parallel

await task
```

then it would make sense. But we don't do that.

**Comment by MithrilMan** [comment]
question: DBreeze states this
> Multi-threaded, ACID compliant, with a solution for deadlocks resolving/elimination, **parallel reads** and synchronized writes/reads.

Did we have code where in a particular method we fetch data from different tables? If yes then having async calls allow us to use Task.WhenAll

If we have no such needs, than go for sync (but then we should remove the whole async chain)

**Comment by dangershony** [comment]
@noescape00 please merge conflicts.

I am for pushing this change, we are likely not going to have an embedded asnc db and if we do the benefits are very minor so we can still just block the thread with little consequence

**Pull request merged**